### DISCUSSION TITLE
Why doesn't covariance work in this case?

### DISCUSSION THREAD

**Discussion by fandrei** [dotnet/csharplang#8541] [General]
This code works.
```
abstract class TreeNodeBase
{
	public virtual TreeNodeBase Parent { get; }
	public virtual IReadOnlyList<TreeNodeBase> Children { get; }
}

class TreeNode : TreeNodeBase
{
	public override TreeNode Parent { get; }
	public override IReadOnlyList<TreeNode> Children { get; }
}
```

This code doesn't compile.
```
interface ITreeNode
{
	ITreeNode Parent { get; }
	IReadOnlyList<ITreeNode> Children { get; }
}

class TreeNodeInt : ITreeNode
{
	public TreeNodeInt Parent { get; }
	public IReadOnlyList<TreeNodeInt> Children { get; }
}
```

Why doesn't it work in the second case?

**Comment by TahirAhmadov** [comment]
The `override` rules are different than `interface` implementation rules.

https://github.com/dotnet/csharplang/discussions/296

**Reply by fandrei** [reply]
@TahirAhmadov what's motivation for this difference? Logically, the relationship between the types is same in both cases.

**Reply by TahirAhmadov** [reply]
There are some subtle differences. I don't recall the details right now, but there were some discussions on the subject in other threads. 

However, it doesn't mean your proposal (and #296) can't work.

I do remember somebody (maybe @HaloFour) stating that the way to achieve this would be silently emitting a bridge method, because at the IL level, making variant method implement the interface method is difficult for some reason.

**Reply by fandrei** [reply]
@TahirAhmadov do you remember where those discussions were? Any key words?

**Reply by TahirAhmadov** [reply]
I found #296 by searching for "interface implementation variance" or something similar. I'm pretty sure there were other discussions on this topic, as well, but I couldn't find them (in the admittedly short amount of time I looked) and I definitely don't remember them. Sorry I couldn't help more.

**Comment by CyrusNajmabadi** [comment]
One is an override (where variance is supported). One is an implementation, where it is not. If you wanted it supported, it would need a proposal and specification for how it worked (either with runtime support, or a stub method emitted).

**Reply by fandrei** [reply]
@CyrusNajmabadi Proposal and spec - sure, why not. But right now, I just asked about motivation of this difference.

**Reply by CyrusNajmabadi** [reply]
Basically, things happen in the reverse direction.  They're different because they're different.  One is an override, one is an implementation.  Covariant returns was supported for overrides, as that's what people were asking for.

If want them to be the same, we need motivation for doing that, and then we need to spec out and determine if that is possible :)

Basically, the answer to most things is: because no one asked for it to be that way.

We don't often intentionally go and make things the same.  We need to compelling scenarios (and people actually asking for it) to motivate all the extra work necessary to make it happen.  Variance on overrides was *heavily* requested (including by *me* :D) over the decades.  While you're the first person i remember asking about variance on implementations.

**Reply by fandrei** [reply]
@CyrusNajmabadi too bad. Frederick Brooks said that conceptual integrity is the most important consideration in system design. Similar things must be used in similar ways, as much as it's possible.
In any case I'm sure that I'm not the first, TahirAhmadov already mentioned another one.

**Reply by CyrusNajmabadi** [reply]
> Frederick Brooks said that conceptual integrity is the most important consideration in system design

He's welcome to his opinion.  That has no bearing on our designs and decision making process.  

> In any case I'm sure that I'm not the first, TahirAhmadov already mentioned another one.

General point being that we had thousands of requests for the first across a broad spectrum of users and businesss.  We do not have that here.  Investing in things that are only benefit a tiny amount of users, which takes away from work that benefits the vast majority is not sensible and we will not be changing to that approach, despite what anyone wants or says is good.

**Reply by CyrusNajmabadi** [reply]
You are welcome to request this and see if anyone wants to pick this up.  But in terms of answering your question as to why we haven't done this, i have nothing better to offer than: almost no one has asked, and our time is better spent delivering value that clearly addresses the needs of our community and ecosystem.

**Reply by HaloFour** [reply]
Wasn't there an open question about this when covariant overloads was being discussed?  Something about accidentally picking up implicit implementations unexpectedly?

**Reply by fandrei** [reply]
> Wasn't there an open question about this when covariant overloads was being discussed? Something about accidentally picking up implicit implementations unexpectedly?

Would be interesting to take a look.

**Reply by fandrei** [reply]
> He's welcome to his opinion. That has no bearing on our designs and decision making process.

Sure, it's your funeral. ¯\_(ツ)_/¯

**Reply by CyrusNajmabadi** [reply]
It's been how we've designed things for 25 years so far. No issues yet.

**Reply by fandrei** [reply]
@CyrusNajmabadi maybe that's why C# lost positions to such languages as Python and TypeScript.

**Reply by En3Tho** [reply]
@fandrei Be on topic. You've been told that there were no such request essentially. Make one. Gather up examples where it's useful. Make a push. You want this after all.

Personally I think all these covariant returns are useless and I wouldn't want language design team to waste time on those. Extensions is much more needed feature than this for example. Some lightweight typing explorations, lifetimes, unions - there are zillions of features more useful than this one. But this my own opinion.

Again, you want this feature. Go for it.

**Reply by HaloFour** [reply]
Python climbed the ranks rapidly specifically because of ML and AI.  Javascript did because it is the de facto language of web browsers.  Nobody cares about the zealous adherence to some kind of design philosophy.  If the language solves problems, it will be used to solve those problems.  It almost doesn't even matter how "good" or "bad" the language is, as evidenced by Javascript's marketshare.

**Reply by fandrei** [reply]
@En3Tho sure. I just need to gather some info about the previous discussions first.

**Reply by fandrei** [reply]
@HaloFour how did it happen that only Python benefited from ML and AI? Apparently, because other languages didn't solve the problems that the developers wanted to solve. This happens when you don't do any improvements unless you have to.
And I didn't even mention JavaScript.

**Reply by CyrusNajmabadi** [reply]
@dnfadmin please stay on topic. There's no evidence that success for any of these groups has come from any sort of adherence to the philosophy you are espousing. 

As has already been mentioned, we add things to the language we think provide substantial value to the ecosystem. If you think this should be fine, make a case. We'll consider it against the several other thousands of things that others are asking for as well. If this beats that stuff out, it will be done.

**Reply by fandrei** [reply]
@CyrusNajmabadi I think that it's common knowledge that you must act proactively to be the leader. You can move this discussion somewhere else if you think that it's offtopic here.

**Reply by CyrusNajmabadi** [reply]
@fandrei this is an open source project with heavy community involvement. We've outlined how things are handled here. Thanks.

**Reply by fandrei** [reply]
@CyrusNajmabadi sure. As a developer of this project, do you remember where previous discussions of this topic were?

**Reply by CyrusNajmabadi** [reply]
Off the top of my head, I do not. Covariant returns were several years back, and took runtime support. So it would be around then.

**Comment by RikkiGibson** [comment]
One workaround here would be to explicitly implement the interface member, then declare an additional member which is "the same, but covariant". This is essentially manually implementing the "stub method" which was suggested in https://github.com/dotnet/csharplang/discussions/8541#discussioncomment-11054676.

```cs
interface ITreeNode
{
	ITreeNode Parent { get; }
	IReadOnlyList<ITreeNode> Children { get; }
}

class TreeNodeInt : ITreeNode
{
	ITreeNode ITreeNode.Parent => Parent;
	public TreeNodeInt Parent { get; }

	IReadOnlyList<ITreeNode> ITreeNode.Children => Children;
	public IReadOnlyList<TreeNodeInt> Children { get; }
}
```

**Reply by colejohnson66** [reply]
This is the same pattern used by IEnumerable implementations:

```csharp
public class MyClass : IEnumerable<SomeType>
{
    IEnumerator IEnumerable.GetEnumerator() =>
        GetEnumerator();
    public IEnumerator<SomeType> GetEnumerator()
    {
        // ...
    }
}
```

Give people the generic enumerator, but for users holding onto a non-generic IEnumerable, just forward the call.

**Comment by fandrei** [comment]
Looks like this feature was already discussed in https://github.com/dotnet/csharplang/issues/49 and added to the proposal, but the part about covariance in interfaces was not implemented. The reason is unclear.

**Reply by CyrusNajmabadi** [reply]
Lack of feedback/motivating scenarios.  Availability of easy workarounds.  As mentioned above, we do our investments on where we think we'll get the most bang/buck :)

**Reply by fandrei** [reply]
@CyrusNajmabadi I didn't find anything about dropping this feature for this reason. Can you?
And I can easily find a number of issues about this topic or similar topics.
And by the way, don't start offtopic again. You decided to ignore best practices - ok I got it, it's your call.

**Reply by HaloFour** [reply]
> You decided to ignore best practices

There's nothing "best practice" about doing a bunch of unnecessary work.

**Reply by CyrusNajmabadi** [reply]
> @CyrusNajmabadi I didn't find anything about dropping this feature for this reason. Can you?

As i said, this is hte reverse way of thinking here (and in .net as a whole).  Our default state is: do not do something. 

A feature needs to be worth it *by a lot* to justify the enormous costs associated with doing *anything*. 

> And I can easily find a number of issues about this topic or similar topics.

Yes.  We have thousands of issues on a wealth of topics.  Our job is the literal task of determining *what*, from that *large set* we should do.  This is literally a "bang/buck" conversation.  If we are investing in things with worse bang/buck we are doing a disservice to our community.

> And by the way, don't start offtopic again.

If something is off topic it will be marked off topic.  This is how we moderate things in this repo.

> You decided to ignore best practices - o

Best practice is investing our limited resources where they will provide the most value to the ecosystem.

**Reply by CyrusNajmabadi** [reply]
@fandrei You've made it clear you want that.  We get that.  You've made it clear you don't agree with our decision making process as well. We also get that.  

As we do not intend on changing our process here *at all* the viable forward path us to get this to be picked up by our normal process of determining which of the thousands of features we have in front of to choose from.  If you (or other do that), then this can happen.  If there's no actual demonstration of the need and value for this, especially given the costs involved, then we will continue to invest in the things that are doing that.

Again, you may not like that this is how we approach things.  But railing against that isn't going to change things.

**Reply by fandrei** [reply]
@CyrusNajmabadi  I said - I got it. You decided to work that way - it's your call. But I don't have to agree with your opinion, right?

Ok, back to business. This feature was considered valuable enough to be included into the proposal for covariant return types. I didn't find any discussions that described why it was not implemented. In any case, it's better to continue discussion in that issue.

**Reply by fandrei** [reply]
@HaloFour 
> There's nothing "best practice" about doing a bunch of unnecessary work.

Cutting corners today often results in much more work later.

**Reply by HaloFour** [reply]
> Cutting corners today often results in much more work later.

Not more work than spending tons of additional effort for the sake of placating some overzealous adherence to some design philosophy.  Language design is necessarily pragmatic, and there already are perfectly good solutions in this space.

**Reply by fandrei** [reply]
@HaloFour in this case, it's very far from "tons of additional effort".
By the way, Fred Brooks was as pragmatic as it gets. And not creating problems for your future self is very pragmatic.

**Reply by HaloFour** [reply]
> in this case, it's very far from "tons of additional effort".

It is tons of effort, especially since the runtime doesn't support it.

> And not creating problems for your future self is very pragmatic.

What problems?  Aside some philosophical noise I don't see any problems blocked by a lack of this specific solution.  Explicit interface implementation already enables you to support a covariant implementation, always has.

**Reply by fandrei** [reply]
@HaloFour auto-generation of stub methods would be perfectly enough.

You can say pretty much same thing about most new features discussed here. It's nearly always possible to solve the problem writing more boilerplate code rather than changing language. It's all the matter of convenience.

**Reply by HaloFour** [reply]
> auto-generation of stub methods would be perfectly enough.

That would make it inconsistent, thus bring about the downfall of C#.

> You can say pretty much same thing about most new features discussed here. It's nearly always possible to solve the problem writing more boilerplate code rather than changing language. It's all the matter of convenience.

Sure, but it's a question of ROI.  If enough benefit is demonstrated that it's worth the massive effort (which is the bare minimum amount of effort), then yes, syntax candy will be added to the language.  But, to paraphrase poorly, too much syntax candy leads to syntax cavities.  Not to mention, there enough enough proposals on the backlog on this repo to represent decades of work, many of them easily worth more than this.

**Reply by fandrei** [reply]
@HaloFour what inconsistency you mean?
And many of them are worth much less than this one, including those that were implemented.
